cram/cram_external.c

```diff
@@ -206,12 +502,15 @@ void cram_slice_hdr_get_coords(cram_block_slice_hdr *h,
  *-----------------------------------------------------------------------------
  * cram_block
  */
-int32_t cram_block_get_content_id(cram_block *b)  { return b->content_id; }
+int32_t cram_block_get_content_id(cram_block *b)  {
+    return b->content_type == CORE ? -1 : b->content_id;
+}
 int32_t cram_block_get_comp_size(cram_block *b)   { return b->comp_size; }
 int32_t cram_block_get_uncomp_size(cram_block *b) { return b->uncomp_size; }
 int32_t cram_block_get_crc32(cram_block *b)       { return b->crc32; }
 void *  cram_block_get_data(cram_block *b)        { return BLOCK_DATA(b); }
 int32_t cram_block_get_size(cram_block *b)        { return BLOCK_SIZE(b); }
+int     cram_block_get_method(cram_block *b)      { return b->orig_method; }
```

Comment:
If exposing this data, you also need to make (a subset of?) [`enum cram_block_method_int`](https://github.com/samtools/htslib/blob/f56be61b91bf/cram/cram_structs.h#L215) public so the value returned can be interpreted.  Otherwise you end up with something like [this](https://github.com/samtools/samtools/blob/c6a428f1f8/cram_size.c#L54), which adds a hidden dependency between the samtools and htslib code.

Reply:
It's not really a dependency between samtools and htslib as they both have a dependency on the CRAM specification, which is where these numerical values are canonically defined.

However agreed it's cleaner to have a reduced enum.

Reply:
The issue of whether `cram_block_expand_method` should be public in htslib is an interesting one.  Maybe, but then again maybe it's an htscodecs thing as it doesn't need to take a cram_block and is a generic utility with just a memory buffer?  It could potentially even just have the original method as some "unknown" to make a best guess effort too (given many formats have a magic number).

I chose a single numeric as I'm using it to look up an array for char and string forms, but agreed maybe it just needs a struct and a couple functions that take that struct and turn it into the char/string forms.  It would be cleaner as we could have eg:

```
typedef struct {
    enum hts_codec_method method;
    int level;    // 1 or 9 for gzip min/max, and 1-9 for bzip2
    int order;    // rans4x8 and Nx16
    int rle;      // rans Nx16 specific
    int pack;
    int stripe;
    int cat;
    int Nway;
} hts_codec_method_t;
```

If we did that though, we'd need to make sure the `method` starts at least with the values used in the CRAM specification, even if it includes additional ones like zstd and bsc (which are definite possibilities for future CRAM, but getting a 3.1 2nd implementation was so hard I rolled back on the idea of including them).